src/server/utils/settings-manager/settings-manager.ts

```diff
@@ -126,23 +126,27 @@ export class SettingsManager {
     return this.timeMonitor.timekeeper;
   }
 
-  getSettings(opts: GetSettingsOptions = {}): Promise<AppSettings> {
-    let currentWork = this.currentWork;
+  handleSettingsTask(task: Promise<void>, opts: GetSettingsOptions = {}): Promise<AppSettings> {
+    const timeoutMs = opts.timeout || this.initialLoadTimeout;
+    if (timeoutMs === 0) {
+      return task.then(() => this.appSettings);
+    }
+    return Promise.race([task, timeout(timeoutMs)])
+      .catch(() => {
+        this.logger.error("Settings load timeout hit, continuing");
+      })
+      .then(() => this.appSettings);
+  }
 
-    // Refresh all clusters
-    currentWork = currentWork.then(() => {
+  getFreshSettings(opts: GetSettingsOptions = {}): Promise<AppSettings> {
+    const task = this.settingsLoaded.then(() => {
       return Promise.all(this.clusterManagers.map(clusterManager => clusterManager.refresh())) as any;
     });
+    return this.handleSettingsTask(task, opts);
+  }
 
-    const timeoutPeriod = opts.timeout || this.initialLoadTimeout;
-    if (timeoutPeriod !== 0) {
-      currentWork = Promise.race([currentWork, timeout(timeoutPeriod)])
-        .catch(e => {
-          this.logger.error("Settings load timeout hit, continuing");
-        });
-    }
-
-    return currentWork.then(() => this.appSettings);
+  getSettings(opts: GetSettingsOptions = {}): Promise<AppSettings> {
+    return this.handleSettingsTask(this.settingsLoaded, opts);
```

Comment:
why settingsLoaded is passed as argument if method could access the promise from the object field itself?

Reply:
Because in different invocation we pass decorated promise: https://github.com/allegro/turnilo/blob/840241e289bfbdf8193184f313e487ba6b1f133d/src/server/utils/settings-manager/settings-manager.ts#L145